File: server/routes/listings.js
const express = require('express');
const Listing = require('../models/Listing');
const { protect } = require('../middleware/auth');

const router = express.Router();

// GET /api/listings — browse with optional filters
router.get('/', async (req, res) => {
    try {
        const { city, bhkType, genderPreference, minRent, maxRent } = req.query;
        const filter = { isActive: true };

        if (city) filter.city = { $regex: city, $options: 'i' };
        if (bhkType) filter.bhkType = bhkType;
        if (genderPreference) filter.genderPreference = genderPreference;
        if (minRent || maxRent) {
            filter.rent = {};
            if (minRent) filter.rent.$gte = Number(minRent);
            if (maxRent) filter.rent.$lte = Number(maxRent);
        }

        const listings = await Listing.find(filter)
            .populate('postedBy', 'name avatar city')
            .sort({ createdAt: -1 });

        res.json(listings);
    } catch (err) {
        res.status(500).json({ message: 'Server error', error: err.message });
    }
});

// GET /api/listings/:id — single listing
router.get('/:id', async (req, res) => {
    try {
        const listing = await Listing.findById(req.params.id)
            .populate('postedBy', 'name avatar bio city phone gender age preferences');
        if (!listing) return res.status(404).json({ message: 'Listing not found' });
        res.json(listing);
    } catch (err) {
        res.status(500).json({ message: 'Server error', error: err.message });
    }
});

// POST /api/listings — create (auth required)
router.post('/', protect, async (req, res) => {
    try {
        const listing = await Listing.create({ ...req.body, postedBy: req.user.id });
        res.status(201).json(listing);
    } catch (err) {
        res.status(500).json({ message: 'Server error', error: err.message });
    }
});

// PUT /api/listings/:id — update (owner only)
router.put('/:id', protect, async (req, res) => {
    try {
        const listing = await Listing.findById(req.params.id);
        if (!listing) return res.status(404).json({ message: 'Listing not found' });
        if (listing.postedBy.toString() !== req.user.id)
            return res.status(403).json({ message: 'Not authorized' });

        const updated = await Listing.findByIdAndUpdate(req.params.id, req.body, { new: true });
        res.json(updated);
    } catch (err) {
        res.status(500).json({ message: 'Server error', error: err.message });
    }
});

// DELETE /api/listings/:id — delete (owner only)
router.delete('/:id', protect, async (req, res) => {
    try {
        const listing = await Listing.findById(req.params.id);
        if (!listing) return res.status(404).json({ message: 'Listing not found' });
        if (listing.postedBy.toString() !== req.user.id)
            return res.status(403).json({ message: 'Not authorized' });

        await listing.deleteOne();
        res.json({ message: 'Listing deleted' });
    } catch (err) {
        res.status(500).json({ message: 'Server error', error: err.message });
    }
});

module.exports = router;

File: server/listings.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session, joinedload
from typing import Optional, List
from database import get_db
import models, schemas
from auth import get_current_user

router = APIRouter(prefix="/api/listings", tags=["listings"])

@router.get("/", response_model=List[schemas.ListingOut])
def get_listings(
    city: Optional[str] = None,
    bhk_type: Optional[str] = None,
    gender_preference: Optional[str] = None,
    min_rent: Optional[float] = None,
    max_rent: Optional[float] = None,
    db: Session = Depends(get_db)
):
    query = db.query(models.Listing).options(joinedload(models.Listing.owner)).filter(
        models.Listing.is_active == True
    )
    if city:
        query = query.filter(models.Listing.city.ilike(f"%{city}%"))
    if bhk_type:
        query = query.filter(models.Listing.bhk_type == bhk_type)
    if gender_preference:
        query = query.filter(models.Listing.gender_preference == gender_preference)
    if min_rent:
        query = query.filter(models.Listing.rent >= min_rent)
    if max_rent:
        query = query.filter(models.Listing.rent <= max_rent)

    return query.order_by(models.Listing.created_at.desc()).all()

@router.get("/{listing_id}", response_model=schemas.ListingOut)
def get_listing(listing_id: str, db: Session = Depends(get_db)):
    listing = db.query(models.Listing).options(joinedload(models.Listing.owner)).filter(
        models.Listing.id == listing_id
    ).first()
    if not listing:
        raise HTTPException(status_code=404, detail="Listing not found")
    return listing

@router.post("/", response_model=schemas.ListingOut, status_code=201)
def create_listing(
    body: schemas.CreateListingRequest,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    listing = models.Listing(**body.dict(), posted_by=current_user.id)
    db.add(listing)
    db.commit()
    db.refresh(listing)
    return listing

@router.put("/{listing_id}", response_model=schemas.ListingOut)
def update_listing(
    listing_id: str,
    body: schemas.UpdateListingRequest,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    listing = db.query(models.Listing).filter(models.Listing.id == listing_id).first()
    if not listing:
        raise HTTPException(status_code=404, detail="Listing not found")
    if listing.posted_by != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")

    for key, value in body.dict(exclude_none=True).items():
        setattr(listing, key, value)

    db.commit()
    db.refresh(listing)
    return listing

@router.delete("/{listing_id}")
def delete_listing(
    listing_id: str,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    listing = db.query(models.Listing).filter(models.Listing.id == listing_id).first()
    if not listing:
        raise HTTPException(status_code=404, detail="Listing not found")
    if listing.posted_by != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")

    db.delete(listing)
    db.commit()
    return {"message": "Listing deleted"}
